gpfit/fit_constraintset.py

```diff
@@ -0,0 +1,160 @@
+            self.mfac = Variable("m_{fac-" + name + "-fit}", 1.0, "-",
+                                 "fit factor")
+
+        if fitdata["ftype"] == "ISMA":
+            # constraint of the form 1 >= c1*u1^exp1*u2^exp2*w^(-alpha) + ....
+            alpha = array([fitdata["a%d" % k] for k in
+                           range(fitdata["K"])])
+            lhs, rhs = 1, NomialArray(monos/(ivar/self.mfac)**alpha).sum(0)
+        elif fitdata["ftype"] == "SMA":
+            # constraint of the form w^alpha >= c1*u1^exp1 + c2*u2^exp2 +....
+            alpha = fitdata["a1"]
+            lhs, rhs = (ivar/self.mfac)**alpha, NomialArray(monos).sum(0)
+        elif fitdata["ftype"] == "MA":
+            # constraint of the form w >= c1*u1^exp1, w >= c2*u2^exp2, ....
+            lhs, rhs = (ivar/self.mfac), NomialArray(monos).T
+
+        if fitdata["K"] == 1:
+            # when possible, return an equality constraint
+            if hasattr(rhs, "shape"):
+                if rhs.ndim > 1:
+                    self.constraint = [(lh == rh) for lh, rh in zip(lhs, rhs)]
+                else:
+                    self.constraint = (lhs == rhs)
+            else:
+                self.constraint = (lhs == rhs)
+        else:
+            if hasattr(rhs, "shape"):
+                if rhs.ndim > 1:
+                    self.constraint = [(lh >= rh) for lh, rh in zip(lhs, rhs)]
+                else:
+                    self.constraint = (lhs >= rhs)
+            else:
+                self.constraint = (lhs >= rhs)
+
+        self.bounds = {}
+        for i, dvar in enumerate(self.dvars):
+            self.bounds[dvar] = [fitdata["lb%d" % i],
+                                 fitdata["ub%d" % i]]
+
+        ConstraintSet.__init__(self, [self.constraint])
+
+    def get_fitdata(self):
+        " return fit data "
+        return self.fitdata
+
+    def get_dataframe(self):
+        " return a pandas DataFrame of fit parameters "
+        import pandas as pd
+        df = pd.DataFrame(self.fitdata.values()).transpose()
+        df.columns = self.fitdata.keys()
+        return df
+
+    def process_result(self, result):
+        """
+        make sure fit result is within bounds of fitted data
+        if data comes from Xfoil and airfoil is provided check against xfoil
+        """
+        super(FitCS, self).process_result(result)
+
+        if amax([abs(result["sensitivities"]["constants"][self.mfac])]) < 1e-5:
+            return None
+
+        if self.airfoil:
+            from .xfoilWrapper import xfoil_comparison
+            cl, re = 0.0, 0.0
+            for dvar in self.dvars:
+                if "Re" in str(dvar):
+                    re = result(dvar)
+                if "C_L" in str(dvar):
+                    cl = result(dvar)
+            cd = result(self.ivar)
+            if not hasattr(cl, "__len__") and hasattr(re, "__len__"):
+                cl = [cl]*len(re)
+            err, cdx = xfoil_comparison(self.airfoil, cl, re, cd)
```

Comment:
whoa, does this call xfoil? I think XfoilFitConstraintSet should be a separate class, then :p